### PR TITLE
[WIP] #1186: Improve tests for circuit drawers

### PR DESCRIPTION
**Pull request by Exferro**
### Summary
This pull request intends to improve testing of circuit drawers as was asked in #1186.

### Details and comments
So far this pull request is my first one, so I do apologize if it is rather messy. I treat it to be WIP for I'm not sure that testing technique suits this project. Moreover, one may add other tests as well. 

#### What have I done
I added a file `test_circuit_drawers` to store all my changes. Then, I introduced two test cases:
 - `TestDrawingMethods` which tests how underlying circuit drawers (`latex`, `latex_source`, `text` and `matplotlib`) perform; 
 - `TestCircuitDrawer` which tests behaviour of `circuit_drawer`.

Now, let me explain how do they work.
##### TestDrawingMethods
The main idea of this test case is to compare outputs produced with circuit drawers to the previously obtained reference ones. Work of circuit drawers is tested on different types of QuantumCircuit (small, medium, large and deep).

To do that, I've introduced four functions which return either small, or medium, or large, or deep QuantumCircuit. Then, I've made **all** underlying circuit drawers to generate reference outputs. All these references are stored in corresponding subdirectories of `references` directory.

Next, I've written tests. Since given a circuit type testing routines are the same for all drawers, I've used the subTests feature of `unittest`. Thus, inside the test I iterate over circuit drawers and create a subTest for each one. Each subTest makes corresponding circuit drawer to draw chosen circuit and then compares the output with the reference. If they are similar, subTest is passed. Otherwise, not.

However, that is not the end. Even for different types of circuit tests are pretty similar (I would say they are identical up to one string). Thus, I've added a MegaTest which iterates over **both** circuit types and circuit drawers and then creates a subTest for each pair. However, since it is rather slow, I have `@unittest.skip`ed it.

Note that `setUp` method is used to create a temporary folder to store all outputs. `tearDown` deletes this folder and thus keeps everything clear.

#### TestCircuitDrawer

I have noticed that `circuit_drawer` does nothing but calling underlying circuit drawers upon different values of input parameters. Thus, this test case has another nature: it doesn't test the correctness of the outputs, instead it checks the correctness of `circuit_drawer` behaviour.  

In particular, it tests handling of the following cases:
 - Correct value of `output` provided;
 - Wrong value of `output` provided;
 - No value of `output` provided.

Also, this test case checks how `interactive=True` is handled.

I've implemented it in the following way. For example, I pass to `circuit_drawer` set of parameters which causes the `_latex_circuit_drawer` to be called. Then, I mock the invoked function such that it does nothing and then check that it was called once. 

I should note that in this case I've also used subTests since testing routines are the same in this case as well.

It seems that I have described all my job, hope it won't be to hard too finish this pull request!

### CODE REVIEW

**Comment by Exferro** [comment]
@diego-plan9 Great thanks for your review and hints, they are extremely useful to bring more project style to the PR.

What about the references: yes, they are indeed regenerated manually. I understand that this has obvious drawback. However, I believe that serious changes in the visualization algorithms are by far more rare than some minor changes. Since proposed testing approach aims at catching the bugs caused with the latter, I hope that introduced overhead wouldn't be critical.

Nevertheless, sorting out how to streamline the process of regenerating would be a great deal! Hope, we will figure it out!

File: test/python/tools/visualization/test_circuit_drawers.py [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+from numpy import pi
+
+import qiskit.tools.visualization._circuit_visualization as _cv
+import qiskit.tools.visualization._matplotlib as _mpl
+
+from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister
+from qiskit.tools.visualization import circuit_drawer
+from qiskit.tools.visualization import VisualizationError
+
+from ...common import QiskitTestCase
```
[the remark this answers is not in the document]
Comment (Exferro) [reply test/python/tools/visualization/test_circuit_drawers.py]: Done!

File: test/python/tools/visualization/test_circuit_drawers.py [hunk 2]
```diff
@@ -0,0 +1,610 @@
...
+    circuit.measure(qr, cr)
+
+    return circuit
+
+
+def _medium_circuit():
+    """Creates a medium-size quantum circuit, consisting of three qubits and most used gates applied
+    to them.
+
+    Returns: an instance of QuantumCircuit.
```
[the remark this answers is not in the document]
Comment (Exferro) [reply test/python/tools/visualization/test_circuit_drawers.py]: Done!

File: test/python/tools/visualization/test_circuit_drawers.py [hunk 3]
```diff
@@ -0,0 +1,610 @@
...
+        self.tmp_dir = os.path.join(_this_directory(), 'tmp')
+        if not os.path.exists(self.tmp_dir):
+            os.makedirs(self.tmp_dir)
+
+    def test_small_circuit(self):
+        """Tests whether outputs of different circuit drawers upon drawing a small circuit equal
+         reference outputs.
+        """
+        # Specify a type of circuit used in this test
+        circuit_type = 'small'
```
[the remark this answers is not in the document]
Comment (Exferro) [reply test/python/tools/visualization/test_circuit_drawers.py]: It's a great idea, I don't know why I was so stuck!

Bodies of functions are indeed equal, so I can move them into separate function!

File: test/python/tools/visualization/test_circuit_drawers.py [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+                self.assertOutputsAreEqual(draw_method,
+                                           test_output + self.extensions.get(draw_method, ''),
+                                           reference_output + self.extensions.get(draw_method, ''))
+
+            # We have built a QuantumCircuit and counters of registers increased, we have reset them
+            # such that consequent tests won't fail
+            _reset_registers()
+
+    @unittest.skip('A test which runs tests for all circuit types inside is skipped.'
+                   'Tests for all circuit types are better to be launched separately.')
```
[the remark this answers is not in the document]
Comment (Exferro) [reply test/python/tools/visualization/test_circuit_drawers.py]: @diego-plan9,  on my MacBook Air 15 it took 1m 15s. I don't know whether it is slow or fast corresponding to Qiskit, so I would appreciate any decision)

File: test/python/tools/visualization/test_circuit_drawers.py [hunk 5]
```diff
@@ -0,0 +1,610 @@
...
+from qiskit.tools.visualization import VisualizationError
+
+from ...common import QiskitTestCase
+
+
+def _this_directory():
+    return os.path.dirname(os.path.abspath(__file__))
+
+
+REFERENCES_DIR = os.path.join(_this_directory(), 'references')
```
[the remark this answers is not in the document]
Comment (Exferro) [reply test/python/tools/visualization/test_circuit_drawers.py]: Done!

File: test/python/tools/visualization/test_circuit_drawers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (diego-plan9) [reply test/python/tools/visualization/test_circuit_drawers.py]: Thanks! Then let's use `@slow_test` indeed :+1:

[the remark this answers is not in the document]
Comment (Exferro) [reply test/python/tools/visualization/test_circuit_drawers.py]: @diego-plan9 I would like to mention, that the time required with this test is actually a sum of times required for other tests. So, the second slowest test takes approximately 30 seconds, does it also requires a `@slow test` decorator?

[the remark this answers is not in the document]
Comment (diego-plan9) [reply test/python/tools/visualization/test_circuit_drawers.py]: It is a bit on the long-ish side indeed, but don't worry too much about it - we can adjust later :+1:

File: test/python/tools/visualization/test_circuit_drawers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (diego-plan9) [reply test/python/tools/visualization/test_circuit_drawers.py]: Missed this comment! A follow-up PR was issued #1652 that updated the version for that dependency (locally, you can just do `pip install -U marshmallow`), that hopefully gets rid of the problem - apologies about that!

**Comment by diego-plan9** [comment]
> Nevertheless, sorting out how to streamline the process of regenerating would be a great deal! Hope, we will figure it out!

Sure - it seems it can be tackled after this PR!

Just for double-checking and without any kind of pressure - is there something blocking your work that we might be able to help with (unanswered comment, a specific part of the PR that seems troublesome, etc)?

**Comment by diego-plan9** [comment]
Haha ok! I think Luciano will be back next week - in the meantime, can you try to [get the CI passing](https://travis-ci.org/Qiskit/qiskit-terra/builds/479041915?utm_source=github_status&utm_medium=notification), as we rely on them to ensure the test pass (and having the "green tick" is actually a requisite for merging a PR)?

There seems to be one [small style warning](https://travis-ci.org/Qiskit/qiskit-terra/jobs/479041916#L701) (that can be solved by appending a space in the right place - you can execute `make lint` in your local machine to replicate that check), and a number of tests seem to be [failing](https://travis-ci.org/Qiskit/qiskit-terra/jobs/479041917#L1729): it seems a bunch of them are due to the ground truth circuit files needing to be regenerated, but there are also a couple of them (`test_correct_output_provided`, `test_interactive`) that fail because `assert_called_once` [is only available in Python 3.6+](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once) - and we also support Python 3.5 (hopefully you can simulate it via other assertions in the `mock` module part of 3.5!).

**Comment by Exferro** [comment]
@diego-plan9 Yes, of course, I will fix style bug and bugs related to Python version.

However, since I am not very known with CI systems, I have a question: the reason of CI failure is that **added in this PR** tests fail? And did I catch it well: you suppose that it is due to obsolete references not representing current state of the drawers? Thanks!

**Comment by Exferro** [comment]
> The CIs (Travis) jobs automatically [clone your PR branch](https://travis-ci.org/Qiskit/qiskit-terra/jobs/479041917#L428) and then [launch the tests](https://travis-ci.org/Qiskit/qiskit-terra/jobs/479041917#L699) (or the linter) under different Python versions and operative systems

Ah! Well, I was overconfidently thinking that the system will not include tests unknown to it. I also supposed that the reason for failures is the **presence of unknown tests** itself. Now, I understand it more clearly, thank you very much :-)

So, I dive into fixing all these test, hope this green tick will appear soon!

**Comment by Exferro** [comment]
@diego-plan9 So far, I have explored that tests failed not only because of obsolete references, but because of different Python versions. At the last `push` I've placed references for Python 3.5 and Python 3.6 into different folders and added handling of the current version inside tests. 

However, I am not sure that it is an appropriate solution: of course, adding a reference for each version can manage this, but what shall we do if there are too many versions? Consequently, a question arises: is it OK, that different versions of Python produce different outputs or it is a ground for yet another issue?

Thanks in advance!

**Comment by diego-plan9** [comment]
Hmm, adding files per version/os would not really scale - can you try another approach? Looking at the logs, it seems to be the `text` references that are failing (based on a quick look at https://travis-ci.org/Qiskit/qiskit-terra/jobs/479987583 - take my comment with caution), and there is a suspicious `q27_0` mention at line L1669: can you add `name`s to the registers and circuits manually during their instantiation, as that might be the only real difference? The register and circuit names are automatically added if none is specified, and they are autoincremented during the same session.

If it is the `assertImagesAreEqual` that is failing, perhaps you can play with the `diff_tolerance` parameter as well - that was actually its original purpose (ie. perform a comparison between images allowing a degree of flexibility to account for small changes). Let's see how it goes!

**Comment by Exferro** [comment]
@diego-plan9 Finally, it came — I mean a green tick! However, I have several notes:

 1. The problem was indeed in naming of registers. I have earlier introduced a function which dropped indices of registers, but its usage was wrong: there were other tests, not only mine. So, I gave all the registers a fixed name — and it worked! Thank you very much for this idea!
 2. There also occurred a problem with `latex_drawer`, it seems like Travis didn't install `pdflatex` — at least, the drawer claimed so. I managed it with handling an `OSError` exception, but it doesn't  like good. Probably, either `latex_drawer` shouldn't "draw" (I mean, we can check the correctness of generated latex source), or `pdflatex` shall be also installed.
 3. The references for 3.5, 3.6 and 3.7 turned out to be enough (there were no system dependency). However, I am not sure that comparing of **output** files is what you want. Probably, it is more fruitful to compare intermediate representations: latex source for both latex drawers; and a `single_string()` representations for `text_drawer` — I guess the difference between versions hides in `dump()` function, not in an algorithm itself.

So, if you find having references for different Python versions not quite appropriate (well, so do I now...), I may go on looking for better testing techniques and continue working on this PR.

Thanks!

**Comment by diego-plan9** [comment]
It's progress! :tada: 

> So, if you find having references for different Python versions not quite appropriate (well, so do I now...), I may go on looking for better testing techniques and continue working on this PR.

Yes, I think this is the main thing that requires some changes. Actually playing a bit based on your latest commit:
```
references$ diff -r 3.6 3.7 -q
references$ diff -r 3.5 3.6/ -q
Files 3.5/deep/latex and 3.6/deep/latex differ
Files 3.5/deep/latex_source and 3.6/deep/latex_source differ
Files 3.5/deep/mpl.png and 3.6/deep/mpl.png differ
Files 3.5/deep/text and 3.6/deep/text differ
```
which seems to imply that all the files for 3.6 and 3.7 are equal, and between 3.5 and 3.6 only the `deep` ones differ. Interestingly, when looking at `3.5/deep/text` and `3.6/deep/text` "visually", they actually seem to represent different circuits (if you check the measures "block", in 3.5 they start at `qr_15`, and in 3.6 they start at `qr_19`). So ... you might have uncovered a bug! We experienced potentially similar issues in the past, I'll look a bit more into it.

The good thing is that, if we assume that it is a bug, we can go simplify things a bit. Can you:
* revert the changes so only one set of references is stored (ie. no per-version files)
* temporarily leave `deep` out of the tests (ie. commenting the entry in `self.circuits`) - this is just while we troubleshoot the bug, and making the CI pass gracefully.

I'll also make a couple of inline comments, and after that hopefully @1ucian0 (who in the end I think will be back next week) or @mtreinish can give a better reply to the remaining questions about the "what we need" and goals for this PR - in a way, continuing the conversation started at #1186. Thanks again!

File: test/python/tools/visualization/test_circuit_drawers.py [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+             bool: True if outputs are similar up to given precision, False otherwise.
+
+        """
+        if draw_method in self.file_output_methods:
+            self.assertFilesAreEqual(test_output, reference_output)
+
+        if draw_method in self.image_output_methods:
+            self.assertImagesAreEqual(test_output, reference_output)
+
+    def assertFilesAreEqual(self, current, expected):
```
[the remark this answers is not in the document]
Comment (Exferro) [reply test/python/tools/visualization/test_circuit_drawers.py]: Done!

**Comment by Exferro** [comment]
@diego-plan9 Thanks for your comments, they are great! So far, I have indeed streamlined the hierarchy of classes, used `tempfile` facilities and made some other cosmetic changes. And, which is more, it now works for all Python version! :-)

Is it now OK, or there are still things to do I haven't noticed? Thanks once again!)

**Comment by Exferro** [comment]
@diego-plan9 🎉

Regarding your distant question, I learnt what to do, I did it and it was done) Now, let's wait!)

File: test/python/tools/visualization/test_circuit_drawers.py [hunk 7]
```diff
@@ -0,0 +1,610 @@
...
+                    # Check that corresponding function was called once
+                    circuit_drawer(None, output=draw_method)
+                    mock_draw_method.assert_called_once()
+
+    def test_wrong_output_provided(self):
+        """Tests correct exceptioning of wrong output format """
+        with self.assertRaises(VisualizationError):
+            circuit_drawer(None, output='wrong_output')
+
+    @unittest.skip('Test for deprecated fallback is skipped ')
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply test/python/tools/visualization/test_circuit_drawers.py]: yes. it can be removed.

[the remark this answers is not in the document]
Comment (Exferro) [reply test/python/tools/visualization/test_circuit_drawers.py]: Done!

**Comment by Exferro** [comment]
@1ucian0, if it is not too impolite from my side, could you please tell if there is something else to do with this pull request?

Thanks!

**Comment by 1ucian0** [comment]
Yes! It's in my list of pending things... sorry for the delay. I'm trying to catch up with backlog :S

**Comment by 1ucian0** [comment]
Sorry for the delay @Exferro , do you mind updating it so it can be merged? Is there any automatic way to regenerate the reference files?

**Comment by Exferro** [comment]
@1ucian0, it's okay!) And, of course I don't mind doing that)

However, I just would like to wonder what do you mean by **automatic**? At the moment, I can uncomment a part of `TestDrawingMethods.setUp` function so that it will generate new references just before proceeding to tests.

However, I am not quite sure that this fits definition of the automatic, as I have to do some manual work. 

Is it indeed okay or shall we come up with a more smart solution? Thanks!

**Comment by 1ucian0** [comment]
Let me clarify. In https://github.com/Qiskit/qiskit-terra/pull/1571/, if you run the test as a script, the reference file is regenerated. See:
 https://github.com/Qiskit/qiskit-terra/blob/76bea9a7a390476ccbd0b81cb6ba87b29eb4ad95/test/python/transpiler/test_mappers.py#L8-L17

This is what I mean with make the metatesting more "similar to each other". Let me know if you have thoughts about it.

**Comment by Exferro** [comment]
@1ucian0, I see now, thanks for clarification. 

Well, I have a piece of code (commented, as I already mentioned), which generates all the references at once. I think, I may easily move it to a function, which will be called as soon as the file is executed as a standalone script. 

How do you find this approach? Thanks!

**Comment by 1ucian0** [comment]
Yes please. Also, run the tests after generating the ground truth again, so we make sure it's working.

**Comment by Exferro** [comment]
@1ucian0, I've added this functionalty, but in quite a different way: the references are generated with static method of a test case in the main function, not during the set up. 

Anyway, looks like it is done!)

**Comment by 1ucian0** [comment]
Thanks a lot @Exferro ! Merging!

**Pull request merged**